tensorboard/util/grpc_util.py

```diff
@@ -51,7 +53,56 @@
 _VERSION_METADATA_KEY = "tensorboard-version"
 
 
-def async_call_with_retries(api_method, request, done_callback, clock=None):
+class AsyncCallFuture:
+    """Encapsulates the future value of a retriable async gRPC request.
+
+    Abstracts over the set of futures returned by a set of gRPC calls
+    comprising a single logical gRPC request with retries.  Communicates
+    to the caller the result or exception resulting from the request.
+
+    Args:
+      completion_event: The constructor should provide a thredding.Event which
+        will be used to communicate when the set of gRPC requests is complete.
+    """
+
+    def __init__(self, completion_event):
+        self._active_grpc_future = None
+        self._active_grpc_future_lock = threading.Lock()
+        self._completion_event = completion_event
+
+    def _set_active_future(self, grpc_future):
+        if grpc_future is None:
+            raise RuntimeError(
+                "_set_active_future invoked with grpc_future=None"
+            )
+        # TODO() maybe add validation logic here that future is not None.
+        with self._active_grpc_future_lock:
+            self._active_grpc_future = grpc_future
+
+    def result(self, timeout):
+        """Analogous to `grpc.Future.result`. Returns the value or exception.
+
+        This method will wait until the full set of gRPC requests is complete
+        and then act as `grpc.Future.result` for the single gRPC invocation
+        corresponding to the first successful call or final failure, as
+        appropriate.
+
+        Args:
+          timeout: How long to wait in seconds before giving up and raising
```

Comment:
As written, this code won't necessarily raise a timeout exception, since it could be the case that we wait for completion and never get there, but then the active future is a completed one from the intermediate future.  Imagine that we have `timeout=0.000001` - then we basically skip the `wait()` on the completion event entirely, and we'll just return whatever the result of the first future was, when it finishes (could be success or failure).

Instead, it might be better to check the return value from `wait()` ([docs](https://docs.python.org/3/library/threading.html#threading.Event.wait)) and then if it's False we raise a "timeout" exception ourselves.  That seems better than exposing the intermediate state of whatever random underlying future happens to be active when the timeout expires.

Reply:
Good catch, thanks.  Added explicit exception throwing and added a test for the case.